Added retry topics (#166)
agentgateway/website: assets/agw-docs/pages/resiliency/retry/retry.md

A request retry is the number of times a request is retried if it fails. This setting can be useful to avoid your apps from failing if they are temporarily unavailable. With retries, calls are retried a certain number of times before they are considered failed. Retries can enhance your app's availability by making sure that calls don't fail permanently because of transient problems, such as a temporarily overloaded service or network.

<!-- TO DO: Is the sample app needed since another is installed? -->
{{< reuse "agw-docs/snippets/agentgateway/prereq.md" >}}
 

## Step 1: Set up your environment

> 1. Install the experimental Kubernetes Gateway API CRDs.

Reviewer: since this is just one step, you can just add it to the steps below.

   | Field | Description |
   |-------|-------------|
   | `hostnames` | The hostnames to match the request, such as `retry.example`. |
   | `parentRefs` | The gateway to which the request is sent. In this example, you select the `agentgateway-proxy` gateway that you set up before you began. |
   | `rules` | The rules to apply to requests. |
   | `matches` | The path to match the request. In this example, you match any requests to the sample app with path prefix `/`. |
   | `path` | The path to match the request. In this example, you match all paths with the prefix `/`. |

>    | `backendRefs` | The backend service to which the request is sent. In this example, you select the `httpbin` service that you set up in the previous step. |

Reviewer: ```suggestion
   | `backendRefs` | The backend service to which the request is sent. In this example, you select the `httpbin` service that you set up in [before you begin](#before-you-begin). |
```

   | `retry.attempts` | The number of times to retry the request. In this example, you retry the request 3 times. |
   | `retry.backoff` | The duration to wait before retrying the request. In this example, you wait 1 second before retrying the request. |
   | `retry.codes` | The HTTP status codes for which the gateway retries the request. In this example, the request is retried if the backend returns 500 or 503. The gateway also retries on connection failures (for example, connection refused). |

2. Verify that the gateway proxy is configured to retry the request.

   1. Port-forward the gateway proxy on port 15000.

      ```sh
      kubectl port-forward deploy/agentgateway-proxy -n {{< reuse "agw-docs/snippets/namespace.md" >}} 15000
      ```

   2. Get the configuration of your gateway proxy as a config dump.

      ```sh
      http://localhost:15000/config_dump
      ```

   3. Find the route configuration for the cluster in the config dump. Verify that the retry policy is set as you configured it.
      
      Example `jq` command:
      
      ```sh

>       curl -s http://localhost:15000/config_dump | jq '.listeners."agentgateway-system/agentgateway-proxy.http".routes | to_entries[] | select(.value.name == "retry" and (.value.inlinePolicies[]? | has("retry"))) | .value'

Reviewer: the curl command does not work for me 

```
curl -s http://localhost:15000/config_dump | jq '.listeners."agentgateway-system/agentgateway-proxy.http".routes | to_entries[] | select(.value.name == "retry" and (.value.inlinePolicies[]? | has("retry"))) | .value'

jq: error (at <stdin>:874): null (null) has no keys
```

              ],
              "inlinePolicies": [
                {
                  "retry": {
                    "attempts": 3,
                    "backoff": "1s",
                    "codes": [
                      "500 Internal Server Error",
                      "503 Service Unavailable",
                    ]
                  }
                }
              ]
            }
          }
        }
      }
      ```
   
   {{% /tab %}}
   {{% tab tabName="HTTPRoute (AgentgatewayPolicy)" %}}
   1. Create an HTTPRoute that routes requests along the `retry.example` domain to the sample app.
      ```yaml
      kubectl apply -f- <<EOF
      apiVersion: gateway.networking.k8s.io/v1
      kind: HTTPRoute
      metadata:
        name: retry
        namespace: httpbin
      spec:
        hostnames:
        - retry.example
        parentRefs:
        - name: agentgateway-proxy
          namespace: {{< reuse "agw-docs/snippets/namespace.md" >}}
        rules:
        - matches: 
          - path:
              type: PathPrefix
              value: /
          backendRefs:
          - name: httpbin
            namespace: httpbin
            port: 8000 
      EOF
      ```

>    2. Create an {{< reuse "agw-docs/snippets/trafficpolicy.md" >}} that applies a retry policy to the HTTPRoute rule. 

Reviewer: in order to apply to an HTTPRoute rule, you must have a name in the HTTPRoute rules section, and then use `sectionName` in the AgentgatewayPolicy targetRefs section. 

https://kgateway.dev/docs/envoy/latest/resiliency/retry/retry/

   ```
   {{% /tab %}}
   {{% tab tabName="Port-forward for local testing" %}}
   ```sh
   curl -vi localhost:8080/status/500 -H "host: retry.example"
   ```
   {{% /tab %}}
   {{< /tabs >}}

   Example output:

   ```
   ...
   < HTTP/1.1 500 Internal Server Error
   ...
   ```

2. Verify that the request was retried. Look for `retry.attempt=3` in the output.

   ```sh
   kubectl logs -n {{< reuse "agw-docs/snippets/namespace.md" >}} \
   -l gateway.networking.k8s.io/gateway-name=agentgateway-proxy \
   --tail=1 | grep -E 'retry.example' 
   ```

   Example output:
   ```
   info	request gateway=agentgateway-system/agentgateway-proxy
   listener=http route=httpbin/retry endpoint=10.244.0.21:8080
   src.addr=127.0.0.1:59284 http.method=GET http.host=retry.example
   http.path=/status/500 http.version=HTTP/1.1 http.status=500
   protocol=http retry.attempt=3 duration=1ms
   ```

## Cleanup

{{< reuse "agw-docs/snippets/cleanup.md" >}}

1. Delete the HTTPRoute resource. 
   
   ```sh
   kubectl delete httproute retry -n httpbin
   ```
2. If you created an {{< reuse "agw-docs/snippets/trafficpolicy.md" >}}, delete it.
   ```sh

>    kubectl delete {{< reuse "agw-docs/snippets/trafficpolicy.md" >}} retry -n httpbin

Author: Fixed

   ```
   {{% /tab %}}
   {{% tab tabName="Port-forward for local testing" %}}
   ```sh
   curl -vi localhost:8080/status/500 -H "host: retry.example"
   ```
   {{% /tab %}}
   {{< /tabs >}}

   Example output:

   ```
   ...
   < HTTP/1.1 500 Internal Server Error
   ...
   ```

2. Verify that the request was retried. Look for `retry.attempt=3` in the output.

   ```sh
   kubectl logs -n {{< reuse "agw-docs/snippets/namespace.md" >}} -l gateway.networking.k8s.io/gateway-name=agentgateway-proxy --tail=1 | grep -E 'retry.example' 
   ```

   Output:
   ```
   info	request gateway=agentgateway-system/agentgateway-proxy listener=http route=httpbin/retry endpoint=10.244.0.21:8080 src.addr=127.0.0.1:59284 http.method=GET http.host=retry.example http.path=/status/500 http.version=HTTP/1.1 http.status=500 protocol=http retry.attempt=3 duration=1ms
   ```

## Cleanup

{{< reuse "agw-docs/snippets/cleanup.md" >}}

1. Delete the HTTPRoute resource. If you used the HTTPRoute (Kubernetes GW API) tab, the route is in the `httpbin` namespace; otherwise delete from both namespaces as needed.
   
   ```sh
   kubectl delete httproute retry -n httpbin
   ```

> 2. If you used the AgentgatewayPolicy or Gateway listener tab, delete the {{< reuse "agw-docs/snippets/trafficpolicy.md" >}}.

Reviewer: ```suggestion
2. If you created an {{< reuse "agw-docs/snippets/trafficpolicy.md" >}}, delete it.
```
